### DISCUSSION TITLE
[Help] runtime node_module loading in PnP mode

### DISCUSSION THREAD

**Discussion by Eomm** [yarnpkg/berry#3369] [Q&A] [answered]
Hi,
I'm writing some unit tests to check if `my-module` loads correctly a user-land module `external-module`:

```
const myMod = require('my-module')

myMod( { load: 'external-module' } )
```

Right now I'm having the error:

```
Your application tried to access transport, but it isn't declared in your dependencies; this makes the require call ambiguous and unsound.  Required package: transport Required by: /Users/mspigolon/workspace/pino/test/  Require stack: - /Users/mspigolon/workspace/pino/test/transport.test.js
```

This is loud and clear because of the test that creates at runtime the test dependency

```
  await symlink(
    join(__dirname, 'fixtures', 'transport'),
    join(__dirname, '..', 'node_modules', 'transport')
  )
```

I cannot find a way to say to `yarn@3.0.1` to do the fallback to the filesystem.
I read about the [PnPApi](https://yarnpkg.com/advanced/pnpapi) but I did not figure out how to use them to load this dependency created at runtime. Could you help me? Is it the right choice or do you see a better way?

Another solution would be to publish the `test-transport-module` and install it in my opinion, but I would avoid it.

My `yarnrc` is quite straightforward:

```
yarnPath: ".yarn/releases/yarn-berry.cjs"
nodeLinker: pnp
pnpMode: loose
```

Thks

**Comment by merceyz** [comment, marked as answer]
You should declare the `fixtures/transport` folder as a dependency using the `link:` protocol and let Yarn deal with the linking instead of doing symlinks manually
https://yarnpkg.com/features/protocols

---

> I cannot find a way to say to yarn@3.0.1 to do the fallback to the filesystem.

You can use `pnpIgnorePatterns` to exclude files from the PnP runtime or if you really want to skip the PnP hook and go directly to the native Node resolution you can.
```js
require.resolve('foo', { plugnplay: false });
```
https://yarnpkg.com/configuration/yarnrc#pnpIgnorePatterns

**Reply by Eomm** [reply]
Works:
```
yarn add --dev transport@link:./test/fixtures/transport
```

I did not use the `pnpIgnorePatterns` because I need the other file's requirement need to be resolved by yarn.

Thanks
